proposal/0000-record-dot-syntax.md

```diff
@@ -12,77 +12,276 @@ the link, and delete this bold sentence.**
+
+addYears :: Class -> Int -> Class
+addYears c n = c{taken.year + n} -- update via op
+
+squareUnits :: Class -> Class
+squareUnits c = c{units & (\x -> x * x)} -- update via function
+
+getResults :: [Class] -> [Status]
+getResults = map (.result) -- section
+
+getTerms :: [Class]  -> [Quarter]
+getTerms = map (.taken.term) -- nested section
+```
+
+A full, rigorous set of examples (as tests) are available in the examples directory of [this repository](https://github.com/ndmitchell/record-dot-preprocessor). Those tests include infix applications, polymorphic data types, interoperation with other extensions and more. They follow the [specifications given earlier](#proposed-change-specification).
 
 ## Effect and Interactions
 
-Detail how the proposed change addresses the original problem raised in the
-motivation.
+**Polymorphic updates:** When enabled, this extension takes the `a{b=c}` syntax and uses it to mean `setField`. The biggest difference a user is likely to experience is that the resulting type of `a{b=c}` is the same as the type `a` - you _cannot_ change the type of the record by updating its fields. The removal of polymorphism is considered essential to preserve decent type inference, and is the only option supported by [the `HasField` proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0158-record-set-field.rst).
 
-Discuss possibly contentious interactions with existing language or compiler
-features.
+**Stealing a.b syntax:** The `a.b` syntax is commonly used in conjunction with the `lens` library, e.g. `expr^.field1.field2`. Treating `a.b` without spaces as a record projection would break such code. The alternatives would be to use a library with a different lens composition operator (e.g. `optics`), introduce an alias in `lens` for `.` (perhaps `%`), write such expressions with spaces, or not enable this extension when also using lenses. While unfortunate, we consider that people who are heavy users of lens don't feel the problems of inadequate records as strongly, so the problems are lessened.
 
+**Enabled extensions:** When `RecordDotSyntax` is enabled it should imply the `NoFieldSelectors` extension and allow duplicate record field labels. It would be possible for `RecordDotSyntax` to imply `DuplicateRecordFields`, but we suspect that if people become comfortable with `RecordDotSyntax` then there will be a desire to remove the `DuplicateRecordFields` extension, so we don't want to build on top of it.
 
 ## Costs and Drawbacks
 
-Give an estimate on development and maintenance costs. List how this effects
-learnability of the language for novice users. Define and list any remaining
-drawbacks that cannot be resolved.
+The implementation of this proposal adds code to the compiler, but not a huge amount. Our [prototype implementation](https://gitlab.haskell.org/shayne-fletcher-da/ghc/commits/record-dot-syntax) shows the essence of the parsing changes, which is the most complex part.
```

Comment:
Your prototype implementation does all the desugaring in the parser. That leads to rather confusing errors that refer to the desugared code. Are you planning to keep it this way in the final implementation or will you extend the actual AST which would require significantly more implementation effort but probably provide better errors.

Reply:
We do not plan to keep it that way in the final implementation. We will extend the AST for exactly the reasons you cite.

Reply:
I would probably strike “which is the most complex part” in that case. Modifying the AST and pushing this through the type checker and everything else seems much more scary to me than modifying the parser. Modifying the parser is still a very important step in proving that this change does not create ambiguities but I don’t think it’s the most important change.

Reply:
Can't speak for @ndmitchell but I think he'll be on board with this comment.

Reply:
I was also under the impression from this section and the Proposed Change Spec that this is basically a change to the lexer/parser and not the AST. Like you say this should be clarified.

Reply:
Will clarify. I guess my remark about desugaring in the parser is once you've done that you've nailed down precisely what all the other stages must do. There is no free choice left. While it might be some difficult code, there's no difficult decisions to be made.